estimagic/tests/visualization/test_comparison_plot.py

```diff
@@ -0,0 +1,404 @@
+from estimagic.visualization.comparison_plot import _add_plot_specs_to_df
+from estimagic.visualization.comparison_plot import _create_bounds_and_rect_widths
+from estimagic.visualization.comparison_plot import _determine_figure_height
+from estimagic.visualization.comparison_plot import _determine_plot_heights
+from estimagic.visualization.comparison_plot import _df_with_all_results
+from estimagic.visualization.comparison_plot import _find_next_lower
+from estimagic.visualization.comparison_plot import _find_next_upper
+from estimagic.visualization.comparison_plot import _flatten_dict
+from estimagic.visualization.comparison_plot import _prep_result_df
+
+# ===========================================================================
+# FIXTURES
+# ===========================================================================
+
+
+@pytest.fixture()
+def minimal_res_dict():
+    groups = ["a", "a", "b", "b", "b", "b", "c"]
+    names = ["u", "i", "a", "e", "n", "r", "t"]
+
+    df1 = pd.DataFrame()
+    df1["group"] = groups
+    df1["name"] = names
+    df1["final_value"] = [1.58, 2.01, 2.73, 1.62, 2.18, 1.75, 2.25]
+
+    df2 = pd.DataFrame()
+    df2["group"] = groups + ["c", "d", "d"]
+    df2["name"] = names + ["x", "v", "l"]
+    df2["final_value"] = [1.48, 1.82, 1.12, 2.15, 1.65, 1.93, 2.39, 1.68, -1.24, -0.95]
+
+    df3 = df1.copy(deep=True)
+    df3["final_value"] += [-0.23, -0.2, -0.11, 0.03, -0.13, -0.21, 0.17]
+
+    df4 = df2.copy(deep=True)
+    df4["final_value"] += [0.4, -0.2, -0.6, -0.0, 0.2, -0.1, 0.1, -0.1, 0.0, -0.3]
+
+    res_dict = {
+        "mod1": {"result_df": df1},
+        "mod2": {"result_df": df2},
+        "mod3": {"result_df": df3},
+        "mod4": {"result_df": df4},
+    }
+    return res_dict
+
+
+@pytest.fixture()
+def res_dict_with_model_class(minimal_res_dict):
+    res_dict = minimal_res_dict
+    res_dict["mod2"]["model_class"] = "large"
+    res_dict["mod4"]["model_class"] = "large"
+    return res_dict
+
+
+@pytest.fixture()
+def res_dict_with_cis(res_dict_with_model_class):
+    res_dict = res_dict_with_model_class
+    diff1 = [0.3, 0.1, 0.2, 0.1, 0.3, 0.4, 0.2]
+    diff2 = [0.1, 0.1, 0.2, 0.3, 0.3, 0.2, 0.4]
+    res_dict["mod1"]["result_df"]["conf_int_upper"] = res_dict["mod1"]["result_df"][
+        "final_value"
+    ] + np.array(diff1)
+    res_dict["mod1"]["result_df"]["conf_int_lower"] = res_dict["mod1"]["result_df"][
+        "final_value"
+    ] - np.array(diff2)
+
+    diff1 += [0.3] * 3
+    diff2 += [0.3] * 3
+    res_dict["mod4"]["result_df"]["conf_int_upper"] = res_dict["mod4"]["result_df"][
+        "final_value"
+    ] + 0.5 * np.array(diff1)
+    res_dict["mod4"]["result_df"]["conf_int_lower"] = res_dict["mod4"]["result_df"][
+        "final_value"
+    ] - 0.5 * np.array(diff2)
+
+    return res_dict
+
+
+@pytest.fixture()
+def df():
+    with open("estimagic/tests/visualization/minimal_expected_df.json", "r") as f:
```

Comment:
Pandas has read_json. no need for context manager.

Reply:
changed in 0846b2d5aa68a1880f930eade981bce6930216aa

Reply:
Good point! Implemented in 0846b2d5aa68a1880f930eade981bce6930216aa.